lib/pos_blockchain.go

```diff
@@ -969,6 +969,43 @@ func (bc *Blockchain) getHighestCommittedBlock() (*BlockNode, int) {
 	return nil, -1
 }
 
+func (bc *Blockchain) GetSafeBlocks() ([]*BlockNode, error) {
+	// First get committed tip.
+	committedTip, idx := bc.getHighestCommittedBlock()
+	if idx == -1 || committedTip == nil {
+		return nil, errors.New("GetSafeBlocks: No committed blocks found")
+	}
+	// Now get all blocks from the committed tip to the best chain tip.
+	safeBlocks := []*BlockNode{committedTip}
+	// TODO: How to determine the highest safe block height efficiently.
+	maxHeightSeen := bc.getMaxHeightSeen()
+	for ii := committedTip.Height + 1; ii < maxHeightSeen+1; ii++ {
+		blockNodes, exists := bc.blockIndexByHeight[ii]
+		if !exists {
+			continue
+		}
+		for _, blockNode := range blockNodes {
+			// TODO: Are there other conditions we should consider?
+			if blockNode.IsValidated() {
+				safeBlocks = append(safeBlocks, blockNode)
+			}
+		}
+	}
+	return safeBlocks, nil
+}
+
+func (bc *Blockchain) getMaxHeightSeen() uint32 {
+	// TODO: How to determine the highest safe block height efficiently. We
+	// may need to track it so we don't have to iterate over all heights.
+	maxHeightSeen := uint32(0)
+	for height := range bc.blockIndexByHeight {
+		if height > maxHeightSeen {
+			maxHeightSeen = height
+		}
+	}
+	return maxHeightSeen
+}
```

Comment:
There is a simple way to make this more efficient. Basically, you don't just want the max height seen, but rather the max height that MAY contain a safe block. A block may only be a safe block if it's parent is in the block index. So, if there's a gap in block at block height anywhere after the committed tip, then all blocks beyond it are orphans.

Something like this should work.... but please check my math here lol
```suggestion
func (bc *Blockchain) getMaxSequentialBlockHeightAfter(startingHeight uint32) uint32 {
	hasBlocksAtCurrentHeight := true
	maxSequentialHeightWithBlocks := startingHeight

	for currentHeight := startingHeight; hasBlocksAtCurrentHeight; currentHeight++ {
		maxSequentialHeightWithBlocks = currentHeight
		_, hasBlocksAtCurrentHeight = range bc.blockIndexByHeight[currentHeight]
	}

	return maxSequentialHeightWithBlocks
}
```